overlord/ifacestate/helpers.go

```diff
@@ -540,3 +564,127 @@ func resolveSnapIDToName(st *state.State, snapID string) (name string, err error
+	return func() { mapper = old }
+}
+
+// CoreSnapdMapper implements InterfaceMapper and makes implicit slots
+// appear to be on "core" while they are on "snapd" in reality.
+type CoreSnapdMapper struct{}
+
+// RemapIncomingSlotRef make slots appear on "snapd" rather than "core".
+func (m *CoreSnapdMapper) RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	if slotRef.Snap == "core" {
+		slotRef.Snap = "snapd"
+		return true
+	}
+	return false
+}
+
+// RemapOutgoingSlotRef make slots appear on "core" rather than "snapd".
+func (m *CoreSnapdMapper) RemapOutgoingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	if slotRef.Snap == "snapd" {
+		slotRef.Snap = "core"
+		return true
+	}
+	return false
+}
+
+// RemapIncomingPlugRef doesn't change the plug in any way.
+func (m *CoreSnapdMapper) RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// RemapOutgoingPlugRef doesn't change the plug in any way.
+func (m *CoreSnapdMapper) RemapOutgoingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return false
+}
+
+// Remapping functions for incoming transfers ({state,wire}=>memory)
+
+// RemapIncomingConnRef potentially re-maps connection reference from an API request or being loaded from state.
+//
+// The operation done by remapIncomingConnRef must be symmetric with remapIncomingConnRef.
+// In practice the pair of functions are used to make "snapd" snap the host of implicit
+// interfaces and connections without altering the state in a backwards incompatible way.
+//
+// Data coming from the state and from API requests is changed so that slots on "core"
+// become slots on "snapd" (but only when "snapd" snap itself is being used). When
+// data is about to hit the state again it is re-mapped back.
+func RemapIncomingConnRef(cref *interfaces.ConnRef) (changed bool) {
+	changedPlug := mapper.RemapIncomingPlugRef(&cref.PlugRef)
+	changedSlot := mapper.RemapIncomingSlotRef(&cref.SlotRef)
+	return changedPlug || changedSlot
+}
+
+// RemapIncomingPlugRef potentially re-maps the snap name of a plug reference.
+func RemapIncomingPlugRef(plugRef *interfaces.PlugRef) (changed bool) {
+	return mapper.RemapIncomingPlugRef(plugRef)
+}
+
+// RemapIncomingSlotRef potentially re-maps the snap name of a slot reference.
+func RemapIncomingSlotRef(slotRef *interfaces.SlotRef) (changed bool) {
+	return mapper.RemapIncomingSlotRef(slotRef)
+}
+
+// Remapping functions for outgoing transfers (memory=>{state,wire})
+
+// RemapOutgoingConnRef potentially re-maps connection reference before being saved to store or returned in an API response.
+//
+// This function is symmetric with RemapIncomingConnRef.
+func RemapOutgoingConnRef(cref *interfaces.ConnRef) (changed bool) {
+	changedPlug := mapper.RemapOutgoingPlugRef(&cref.PlugRef)
+	changedSlot := mapper.RemapOutgoingSlotRef(&cref.SlotRef)
+	return changedPlug || changedSlot
+}
+
+// RemapOutgoingPlugRef potentially re-maps the snap name of a plug reference.
```

Comment:
given that snapd will have only  implicit slots afawk,  remapping of plugs is only for completeness but is not stricly needed right?